File: src/logic/account.rs
use crate::account::{BareAccount, EditAccount};
use crate::{database, hasher};
use crate::logic::AuthContext;
use crate::logic::auth::logout;
use crate::database::Error;

use axum::extract::State;
use axum::response::IntoResponse;
use axum::Json;
use http::StatusCode;
use serde::Deserialize;
use sqlx::SqlitePool;

#[derive(Deserialize)]
pub struct PasswordValidation {
    pub current_password: String,
}

#[derive(Deserialize)]
pub struct EditPassword {
    pub current_password: String,
    pub new_password: String,
}

/// Returns the current logged in account object
pub async fn get_account(auth: AuthContext) -> impl IntoResponse {
    (StatusCode::OK, Json(auth.current_user.unwrap())).into_response()
}

/// Creates a new account and returns the just created account object
pub async fn post_account(
    State(pool): State<SqlitePool>,
    Json(payload): Json<BareAccount>,
) -> impl IntoResponse {
    // if username already exists, return with error
    if database::user::exists(pool.clone(), &payload.username).await {
        return (StatusCode::CONFLICT).into_response();
    }

    // create a new user
    let user = match database::account::insert(pool, &payload).await {
        Ok(user) => user,
        Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
    };

    (StatusCode::CREATED, Json(user)).into_response()
}

/// Edit the current logged in account
pub async fn edit_account(
    State(pool): State<SqlitePool>,
    auth: AuthContext,
    Json(payload): Json<EditAccount>,
) -> impl IntoResponse {
    // if username already exists, return with error
    if database::user::exists(pool.clone(), &payload.username).await {
        return (StatusCode::CONFLICT).into_response();
    }

    // edit the accounts information's from the database
    let updated_user =
        match database::account::update(pool, auth.current_user.unwrap().id, &payload).await {
            Ok(user) => user,
            Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
        };

    (StatusCode::OK, Json(updated_user)).into_response()
}

/// Permanently delete the current logged in account
pub async fn delete_account(
    State(pool): State<SqlitePool>,
    auth: AuthContext,
    Json(payload): Json<PasswordValidation>,
) -> impl IntoResponse {
    let account = match database::account::get_id(pool.clone(), auth.clone().current_user.unwrap().id).await {
        Ok(user) => user,
        Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
    };

    if !hasher::verify(&account.password_hash, &payload.current_password) {
        return (StatusCode::UNAUTHORIZED).into_response();
    }

    // delete the account from the database
    let account =
        match database::account::delete(pool, auth.clone().current_user.unwrap().id).await {
            Ok(user) => user,
            Err(Error::SQLX(e)) => return (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
            Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
        };

    // after deletion log out the session
    logout(auth).await;

    (StatusCode::OK, Json(account)).into_response()
}

/// Change the password of the current logged in account
pub async fn edit_account_password(
    State(pool): State<SqlitePool>,
    auth: AuthContext,
    Json(payload): Json<EditPassword>,
) -> impl IntoResponse {
    let account = match database::account::get_id(pool.clone(), auth.clone().current_user.unwrap().id).await {
        Ok(user) => user,
        Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
    };

    if !hasher::verify(&account.password_hash, &payload.current_password) {
        return (StatusCode::UNAUTHORIZED).into_response();
    }

    // delete the account from the database
    let account =
        match database::account::update_password(pool, auth.clone().current_user.unwrap().id, payload.new_password).await {
            Ok(user) => user,
            Err(_) => return (StatusCode::INTERNAL_SERVER_ERROR).into_response(),
        };

    // after deletion log out the session
    logout(auth).await;

    (StatusCode::OK, Json(account)).into_response()
}

File: public/scripts/api/account.js
import {BASE_ACCOUNT_URL, STATUS, Result} from "./main.js"

export class Account {
    constructor(id, username, password_hash) {
        this.id = id;
        this.username = username;
        this.password_hash = password_hash;
    }
}

export class NewAccount {
    constructor(username, password) {
        this.username = username;
        this.password = password;
    }
}

export class EditAccount {
    constructor(username) {
        this.username = username;
    }
}

export class EditPassword {
    constructor(current_password, new_password) {
        this.current_password = current_password;
        this.new_password = new_password;
    }
}

export class PasswordValidation {
    constructor(current_password) {
        this.current_password = current_password;
    }
}

/// returns the current account
/// :return `Account`
export async function get() {
    const request = new Request(`${BASE_ACCOUNT_URL}`, {
        method: "GET",
        credentials: 'include',
    });

    let response = await fetch(request);
    if (response.status === STATUS.OK) {
        let value = await response.json();
        return new Result(true, new Account(value.id, value.username, value.password_hash));
    } else {
        let error = await response.text();
        return new Result(false, error);
    }
}

/// create a new account
/// :param account `NewAccount`
/// :return `Account` newly created account
export async function create(account) {
    const request = new Request(`${BASE_ACCOUNT_URL}`, {
        method: "POST",
        headers: new Headers({
            "Content-Type": "application/json",
        }),
        body: JSON.stringify(account),
        credentials: 'include',
    });

    let response = await fetch(request);
    if (response.status === STATUS.CREATED) {
        let value = await response.json();
        return new Result(true, new Account(value.id, value.username, value.password_hash));
    } else {
        let error = await response.text();
        return new Result(false, error);
    }
}

/// edit an account
/// :param account `EditAccount`
/// :return `Account` edited account
export async function edit(account) {
    const request = new Request(`${BASE_ACCOUNT_URL}`, {
        method: "PUT",
        headers: new Headers({
            "Content-Type": "application/json",
        }),
        body: JSON.stringify(account),
        credentials: 'include',
    });

    let response = await fetch(request);
    if (response.status === STATUS.OK) {
        let value = await response.json();
        return new Result(true, new Account(value.id, value.username, value.password_hash));
    } else {
        let error = await response.text();
        return new Result(false, error);
    }
}

/// edit the password of an account
/// :param current_password the accounts current password
/// :param new_password the new password for the account
/// :return `Account` updated account
export async function edit_password(edit_password) {
    const request = new Request(`${BASE_ACCOUNT_URL}/password`, {
        method: "PUT",
        headers: new Headers({
            "Content-Type": "application/json",
        }),
        body: JSON.stringify({
            current_password: edit_password.current_password,
            new_password: edit_password.new_password
        }),
        credentials: 'include',
    });

    let response = await fetch(request);
    if (response.status === STATUS.OK) {
        let value = await response.json();
        return new Result(true, new Account(value.id, value.username, value.password_hash));
    } else {
        let error = await response.text();
        return new Result(false, error);
    }
}

/// deletes the account
/// :param current_password the accounts current password
/// :return `Account` deleted account
export async function remove(current_password) {
    const request = new Request(`${BASE_ACCOUNT_URL}`, {
        method: "DELETE",
        headers: new Headers({
            "Content-Type": "application/json",
        }),
        body: JSON.stringify({current_password: current_password}),
        credentials: 'include',
    });

    let response = await fetch(request);
    if (response.status === STATUS.OK) {
        let value = await response.json();
        return new Result(true, new Account(value.id, value.username, value.password_hash));
    } else {
        let error = await response.text();
        return new Result(false, error);
    }
}
